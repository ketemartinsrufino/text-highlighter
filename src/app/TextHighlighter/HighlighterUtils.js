const createNewNode = (color) => {
  var newNode = document.createElement("span");
  newNode.setAttribute(
      "style",
      `background-color: ${color}; display: inline;` 
  );
  return newNode;
}

const highlightRange = (range, text, color) => {
  range.surroundContents(createNewNode(color));
  return { text, color };
}

export const handleMouseUp = (color) => { 
  const userSelection = window.getSelection();
  const range = userSelection.getRangeAt(0);
  const text = userSelection.toString();

  return text ? highlightRange(range, text, color) : null;
}