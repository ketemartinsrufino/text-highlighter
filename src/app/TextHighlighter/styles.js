import styled from 'styled-components';

export const TextContainerStyled = styled.div`
  width: 99%;
  height: 100px;
  border: 1px #ccc solid;
    padding: 3px;
`