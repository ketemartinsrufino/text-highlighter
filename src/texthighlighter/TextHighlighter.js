import React, { Component } from 'react';
import styled from 'styled-components';
import ColorList from '../colorList/ColorList';

const TextHighlighterStyled = styled.div`
    border: 1px red solid;
`;

class TextHighlighter extends Component {
    constructor(props) {
        super(props);
        this.state = {
            selectedColor: '',
            selectedTexts: {}
        };
    }
    
    onColorSelect = (color) => {
        this.setState({selectedColor: color})
    }

    onSelectText = () => {
        const color = this.state.selectedColor;
        if(!color) {
            alert("Select a color, please");
            return;
        }
        const highlightRange = (selection) => {
            const range = selection.getRangeAt(0);
            const text = selection.toString();
            var newNode = document.createElement("span");
            newNode.setAttribute(
               "style",
               `background-color: ${color}; display: inline;` 
            );
            range.surroundContents(newNode);
            const selectedByColor = this.state.selectedTexts[color] || [];

            selectedByColor.push(text);
            this.setState((state) => {
                return {
                    ...state,
                    selectedTexts: {
                        ...state.selectedTexts, 
                        [color]: selectedByColor
                    }
                }
            });
        }

        const userSelection = window.getSelection();
        highlightRange(userSelection);
    }

    render(){
        return (
            <TextHighlighterStyled>
                <ColorList onColorSelect={this.onColorSelect}/>
                <div onMouseUp={this.onSelectText}>{this.props.children}</div>
            </TextHighlighterStyled>
        )
    }
}

export default TextHighlighter;