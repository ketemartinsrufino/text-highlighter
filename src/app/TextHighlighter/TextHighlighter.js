import React, { Component } from 'react';
import { HighlighterContext } from '../../context/HighlighterContext';
import { ContainerStyled } from '../styles';
import ColorListSelectOne from '../../components/ColorListSelectOne/ColorListSelectOne';
import TextArea from '../../components/TextArea/TextArea';
import { TextContainerStyled } from './styles';
import EditButton from './components/EditButton';
import { handleMouseUp } from './HighlighterUtils';

class TextHighlighter extends Component {
  constructor(props) {
    super(props);
  }

  onColorSelect = (color) => {
    this.context.updateContext('colorToHighlight', color);
  }

  onSelectText = () => {
    const color = this.context.colorToHighlight;
    if( !color ) {
      alert("Select a color, please");
      return;
    }
    this.context.addHighlightedText(handleMouseUp(color));
  }

  updateText = (event) => {
    this.context.updateContext('originalText', event.target.value);
  }

  enableEditionMode = () => {
    this.context.toogleEditionMode(true);
    this.context.clearHighlightedTexts();
  }
  onBlur = () => {
    this.context.toogleEditionMode(false);
  }

  getContent = () => {
    const text = this.context.originalText;
    const isEditionMode = this.context.isEditionMode
    if( isEditionMode ) {
      return (
        <TextArea  
          value={text} 
          onChange={this.updateText}
          onBlur={this.onBlur}
        />
      )
    } else {
      return (
        <TextContainerStyled onMouseUp={this.onSelectText}>
          {text}   
        </TextContainerStyled>
      )
    }
  }

  render(){
    return (
      <ContainerStyled>
        <ColorListSelectOne 
          selectedColor={this.context.colorToHighlight}
          onColorSelect={this.onColorSelect}/>
        
        {this.getContent()}
        
        <EditButton 
          isEnabled={!this.context.isEditionMode}
          onClick={this.enableEditionMode}
        />
      </ContainerStyled>
    )
  }
}

TextHighlighter.contextType = HighlighterContext;

export default TextHighlighter;