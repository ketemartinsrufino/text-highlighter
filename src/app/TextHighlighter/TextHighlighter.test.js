import React from 'react';
import { shallow } from 'enzyme';
import TextHighlighter from './TextHighlighter';
import toJson from 'enzyme-to-json';

describe('TextHighlighter', () => {
  it('rendering with default values', () => {
    const component = shallow(<TextHighlighter/>);
    expect(toJson(component)).toMatchSnapshot();
  });
  
  it('rendering with textValue values', () => {
    const component = shallow(<TextHighlighter/>);
    expect(toJson(component)).toMatchSnapshot();
  });
});