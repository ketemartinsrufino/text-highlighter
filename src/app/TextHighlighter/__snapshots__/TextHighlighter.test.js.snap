// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextHighlighter rendering with default values 1`] = `
<styled.div>
  <ColorListSelectOne
    onColorSelect={[Function]}
  />
  <styled.div
    onMouseUp={[Function]}
  />
  <EditButton
    isEnabled={true}
    onClick={[Function]}
  />
</styled.div>
`;

exports[`TextHighlighter rendering with textValue values 1`] = `
<styled.div>
  <ColorListSelectOne
    onColorSelect={[Function]}
  />
  <styled.div
    onMouseUp={[Function]}
  />
  <EditButton
    isEnabled={true}
    onClick={[Function]}
  />
</styled.div>
`;
